File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/InventoryQuery/Filters/AvailabilityFilterDecorator.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.InventoryQuery.Filters;

/// <summary>
///     A filter that filters inventory items by their availability.
/// </summary>
public sealed class AvailabilityFilterDecorator : InventoryQueryDecoratorBase
{
    private readonly bool? _isAvailable;

    /// <summary>
    ///     Initializes a new instance of the <see cref="AvailabilityFilterDecorator" /> class.
    /// </summary>
    /// <param name="inner">The inner query to decorate.</param>
    /// <param name="isAvailable">
    ///     True to filter for in-stock items, false for out-of-stock, or null for no availability
    ///     filtering.
    /// </param>
    public AvailabilityFilterDecorator(IInventoryQuery inner, bool? isAvailable) : base(inner)
    {
        _isAvailable = isAvailable;
    }

    /// <inheritdoc />
    protected override string? AppliedDescription
    {
        get
        {
            if (_isAvailable is null) return null;
            return _isAvailable.Value
                ? "Filter: Availability = In Stock (Quantity > 0)"
                : "Filter: Availability = Out of Stock (Quantity = 0)";
        }
    }

    /// <inheritdoc />
    protected override IReadOnlyList<InventoryItem> Decorate(IReadOnlyList<InventoryItem> items)
    {
        if (_isAvailable is null) return items;

        var filtered = _isAvailable.Value ? items.Where(i => i.IsAvailable) : items.Where(i => !i.IsAvailable);
        return filtered.ToList().AsReadOnly();
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventoryquery/filters/AvailabilityFilterDecorator.java
package assignment2solution.domain.inventoryquery.filters;

import assignment2solution.domain.inventory.InventoryItem;
import assignment2solution.domain.inventoryquery.IInventoryQuery;
import assignment2solution.domain.inventoryquery.InventoryQueryDecoratorBase;

import java.util.List;
import java.util.stream.Collectors;

/**
 * A filter that filters inventory items by their availability.
 */
public final class AvailabilityFilterDecorator extends InventoryQueryDecoratorBase {
    private final Boolean isAvailable;

    /**
     * Initializes a new instance of the {@link AvailabilityFilterDecorator} class.
     */
    public AvailabilityFilterDecorator(IInventoryQuery inner, Boolean isAvailable) {
        super(inner);
        this.isAvailable = isAvailable;
    }

    @Override
    protected String getAppliedDescription() {
        if (isAvailable == null) {
            return null;
        }
        return isAvailable
            ? "Filter: Availability = In Stock (Quantity > 0)"
            : "Filter: Availability = Out of Stock (Quantity = 0)";
    }

    @Override
    protected List<InventoryItem> decorate(List<InventoryItem> items) {
        if (isAvailable == null) {
            return items;
        }

        var filtered = isAvailable ? items.stream().filter(InventoryItem::isAvailable)
            : items.stream().filter(i -> !i.isAvailable());

        return filtered.collect(Collectors.toUnmodifiableList());
    }
}
